File: kb_languages/python/knowledge_base/kb_python/kb_memory/search_mem_db.py
from basic_contruct_db import BasicConstructDB


class SearchMemDB(BasicConstructDB):
    def __init__(self,host,port,dbname,user,password,table_name):
        BasicConstructDB.__init__(self,host,port,dbname,user,password,table_name)
        
        self.import_from_postgres(self.table_name)
        self.keys = self._generated_decoded_keys(self.data)
        
        
    def _generated_decoded_keys(self,data):
        self.kbs = {}
        self.labels = {}
        self.names = {}
        self.decoded_keys = {}
        for key in data.keys():
            self.decoded_keys[key] = key.split(".")
            kb = self.decoded_keys[key][0]
            label = self.decoded_keys[key][-2]
            name =  self.decoded_keys[key][-1]
            if kb not in self.kbs:
                self.kbs[kb] = []
            self.kbs[kb].append(key)
            if label not in self.labels:
                self.labels[label] = []
            self.labels[label].append(key)
            if name not in self.names:
                self.names[name] = []
            self.names[name].append(key)
 
        
    def clear_filters(self):
        """
        Clear all filters and reset the query state.
        """
        self.filter_results = self.data
        
        
        
    def search_kb(self, knowledge_base):
        """
        Search for rows matching the specified knowledge_base.
        
        Args:
            knowledge_base: The knowledge_base name to search for.
        """
        new_filter_results = {}
        for key in self.filter_results.keys():
            if key in self.kbs[knowledge_base]:
            
                new_filter_results[key] = self.filter_results[key]
        self.filter_results = new_filter_results
    
        return self.filter_results
        
    def search_label(self, label):
        """
        Search for rows matching the specified label.
        
        Args:
            label: The label value to match
        """
        new_filter_results = {}
        for key in self.filter_results.keys():
            if key in self.labels[label]:
                new_filter_results[key] = self.filter_results[key]
        self.filter_results = new_filter_results
        return self.filter_results
    
    def search_name(self, name):
        """
        Add a filter to search for rows matching the specified name.
        
        Args:
            name: The name value to match
        """
        new_filter_results = {}
        for key in self.filter_results.keys():
            if key in self.names[name]:
                new_filter_results[key] = self.filter_results[key]
        self.filter_results = new_filter_results
        return self.filter_results
    
    def search_property_key(self, data_key):
        new_filter_results = {}
        for key in self.filter_results.keys():
            value = self.data[key].data
            
            if data_key in value:
                new_filter_results[key] = self.filter_results[key]
        self.filter_results = new_filter_results
        return self.filter_results
    
    def search_property_value(self, data_key, data_value):
        """
        Add a filter to search for rows where the properties JSON field contains 
        the specified key with the specified value.
        
        Args:
            key: The JSON key to search
            value: The value to match for the key
        """
        new_filter_results = {}
        for key in self.filter_results.keys():
            value = self.data[key].data
            if data_key in value:
                 if data_value == value[data_key]:
                    new_filter_results[key] = self.filter_results[key]
        self.filter_results = new_filter_results
        return self.filter_results
    



    def search_starting_path(self, starting_path):
        """
        Version that returns filtered results without modifying the original filter_results.
        """
      
        if not isinstance(starting_path, str):
            raise ValueError("starting_path must be a string")
        
        new_filter_results = {}
        
        # Add starting path if it exists
        if starting_path in self.filter_results:
            new_filter_results[starting_path] = self.filter_results[starting_path]
        else:
            self.filter_results = {}
        
        # Get and add descendants
        try:
            starting_path_list = self.query_descendants(starting_path)
            if starting_path_list:
                for item in starting_path_list:
                    if item['path'] in self.filter_results:
                        new_filter_results[item['path']] = self.filter_results[item['path']]
        except Exception as e:
            print(f"Error querying descendants: {e}")
        
        self.filter_results = new_filter_results
        return new_filter_results
            
    def search_path(self,operator,starting_path):
        """
        Add a filter to search for rows matching the specified LTREE path expression.
        
        Args:
            operator: '@>', '<@', '~', '@@', '||'
           starting_path: The LTREE path expression to match
                Examples:
                - 'docs.technical' for exact match
                - 'docs.*' for all immediate children of docs
                - 'docs.*{1,3}' for children up to 3 levels deep
                - 'docs.technical.*' for all children of docs.technical
                - '*.technical.*' for any path containing 'technical'
        """
        search_results = self.query_by_operator(operator,starting_path)
        
        new_filter_results = {}
        for item in search_results:
            if item['path'] in self.filter_results:
                new_filter_results[item['path']] = self.filter_results[item['path']]
        self.filter_results = new_filter_results
        return self.filter_results
    

    def find_descriptions(self, key):
        """
        Extract description from properties field of query results.
        
        Args:
            key_data: Single row dictionary or list of row dictionaries
            
        Returns:
            list: List of dictionaries mapping path to description
        """
        if isinstance(key, list):
            key_data = key
        else:
            key_data = [key]
        
        return_values = {}
        
        for row_key in self.data.keys():
            row_data = self.data[row_key]
            data = row_data.data
            description = data.get('description', '') 
            return_values[row_key] = description
        return return_values
   
if __name__ == "__main__":
    password = input("Enter Postgres Password: ")
    kb = SearchMemDB(host="localhost", port=5432, dbname="knowledge_base", user="gedgar", password=password, table_name='composite_memory_kb')
    print("decoded_keys keys: ",kb.decoded_keys.keys())
    kb.clear_filters()
    print("----------------------------------")
    kb.search_kb("kb1")
    print("----------------------------------")
    print("search kb: ",kb.filter_results.keys())
    print("----------------------------------")
    kb.search_label("info1_link")
    print("search label: ",kb.filter_results.keys())
    kb.search_name("info1_name")
    print("----------------------------------")
    print("search name: ",kb.filter_results.keys())
    
    kb.clear_filters()
    print("search_property_key",kb.search_property_value('data','info1_data'))
    print("----------------------------------")
    kb.clear_filters()
    print("search property key: ",kb.search_property_key("data"))
    print("----------------------------------")
    
    kb.clear_filters()
    print("search_starting_path",kb.search_starting_path("kb2.header2_link.header2_name"))
    print("----------------------------------")
    kb.clear_filters()
    print("search_path",kb.search_path("~","kb2.**"))
    print("----------------------------------")
    kb.clear_filters()
    print("find_descriptions",kb.find_descriptions("kb2.header2_link.header2_name"))
    print("----------------------------------")
    
    

File: building_blocks/python/libraries/file_loader/test_files/kb_go/kb_memory/kb_memory_module/SearchMemDB.go
package kb_memory_module

import (
	"fmt"
	//"log"
	"strings"
)

// SearchMemDB extends BasicConstructDB with search and filtering capabilities
type SearchMemDB struct {
	*BasicConstructDB                    // Embedded struct for inheritance-like behavior
	keys            map[string][]string  // Generated decoded keys
	kbs             map[string][]string  // Knowledge bases mapping
	labels          map[string][]string  // Labels mapping
	names           map[string][]string  // Names mapping
	DecodedKeys     map[string][]string  // Decoded path keys
	FilterResults   map[string]*TreeNode // Current filter results
}

// NewSearchMemDB creates a new SearchMemDB instance and loads data from PostgreSQL
func NewSearchMemDB(host string, port int, dbname, user, password, tableName string) (*SearchMemDB, error) {
	smdb := &SearchMemDB{
		BasicConstructDB: NewBasicConstructDB(host, port, dbname, user, password, tableName),
		kbs:              make(map[string][]string),
		labels:           make(map[string][]string),
		names:            make(map[string][]string),
		DecodedKeys:      make(map[string][]string),
		FilterResults:    make(map[string]*TreeNode),
	}

	// Import data from PostgreSQL
	_, err := smdb.ImportFromPostgres(tableName, "path", "data", "created_at", "updated_at")
	if err != nil {
		return nil, fmt.Errorf("failed to import from postgres: %w", err)
	}

	// Generate decoded keys
	smdb.keys = smdb.generateDecodedKeys(smdb.data)
	
	// Initialize filter results with all data
	smdb.ClearFilters()

	return smdb, nil
}

// generateDecodedKeys processes the data and creates lookup maps
func (smdb *SearchMemDB) generateDecodedKeys(data map[string]*TreeNode) map[string][]string {
	smdb.kbs = make(map[string][]string)
	smdb.labels = make(map[string][]string)
	smdb.names = make(map[string][]string)
	smdb.DecodedKeys = make(map[string][]string)

	for key := range data {
		// Split the key into components
		smdb.DecodedKeys[key] = strings.Split(key, ".")
		
		if len(smdb.DecodedKeys[key]) < 3 {
			// Skip keys that don't have at least kb.label.name structure
			continue
		}

		kb := smdb.DecodedKeys[key][0]
		label := smdb.DecodedKeys[key][len(smdb.DecodedKeys[key])-2]
		name := smdb.DecodedKeys[key][len(smdb.DecodedKeys[key])-1]

		// Add to knowledge bases map
		if _, exists := smdb.kbs[kb]; !exists {
			smdb.kbs[kb] = make([]string, 0)
		}
		smdb.kbs[kb] = append(smdb.kbs[kb], key)

		// Add to labels map
		if _, exists := smdb.labels[label]; !exists {
			smdb.labels[label] = make([]string, 0)
		}
		smdb.labels[label] = append(smdb.labels[label], key)

		// Add to names map
		if _, exists := smdb.names[name]; !exists {
			smdb.names[name] = make([]string, 0)
		}
		smdb.names[name] = append(smdb.names[name], key)
	}

	return smdb.DecodedKeys
}

// ClearFilters clears all filters and resets the query state
func (smdb *SearchMemDB) ClearFilters() {
	smdb.FilterResults = make(map[string]*TreeNode)
	// Copy all data to filter results
	for key, value := range smdb.data {
		smdb.FilterResults[key] = value
	}
}

// SearchKB searches for rows matching the specified knowledge base
func (smdb *SearchMemDB) SearchKB(knowledgeBase string) map[string]*TreeNode {
	newFilterResults := make(map[string]*TreeNode)
	
	if kbKeys, exists := smdb.kbs[knowledgeBase]; exists {
		for _, key := range kbKeys {
			if _, exists := smdb.FilterResults[key]; exists {
				newFilterResults[key] = smdb.FilterResults[key]
			}
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// SearchLabel searches for rows matching the specified label
func (smdb *SearchMemDB) SearchLabel(label string) map[string]*TreeNode {
	newFilterResults := make(map[string]*TreeNode)
	
	if labelKeys, exists := smdb.labels[label]; exists {
		for _, key := range labelKeys {
			if _, exists := smdb.FilterResults[key]; exists {
				newFilterResults[key] = smdb.FilterResults[key]
			}
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// SearchName searches for rows matching the specified name
func (smdb *SearchMemDB) SearchName(name string) map[string]*TreeNode {
	newFilterResults := make(map[string]*TreeNode)
	
	if nameKeys, exists := smdb.names[name]; exists {
		for _, key := range nameKeys {
			if _, exists := smdb.FilterResults[key]; exists {
				newFilterResults[key] = smdb.FilterResults[key]
			}
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// SearchPropertyKey searches for rows that contain the specified property key
func (smdb *SearchMemDB) SearchPropertyKey(dataKey string) map[string]*TreeNode {
	newFilterResults := make(map[string]*TreeNode)
	
	for key := range smdb.FilterResults {
		if node, exists := smdb.data[key]; exists {
			if dataMap, ok := node.Data.(map[string]interface{}); ok {
				if _, hasKey := dataMap[dataKey]; hasKey {
					newFilterResults[key] = smdb.FilterResults[key]
				}
			}
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// SearchPropertyValue searches for rows where the properties JSON field contains the specified key with the specified value
func (smdb *SearchMemDB) SearchPropertyValue(dataKey string, dataValue interface{}) map[string]*TreeNode {
	newFilterResults := make(map[string]*TreeNode)
	
	for key := range smdb.FilterResults {
		if node, exists := smdb.data[key]; exists {
			if dataMap, ok := node.Data.(map[string]interface{}); ok {
				if value, hasKey := dataMap[dataKey]; hasKey {
					if value == dataValue {
						newFilterResults[key] = smdb.FilterResults[key]
					}
				}
			}
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// SearchStartingPath searches for a specific path and all its descendants
func (smdb *SearchMemDB) SearchStartingPath(startingPath string) (map[string]*TreeNode, error) {
	newFilterResults := make(map[string]*TreeNode)
	
	// Add starting path if it exists in filter results
	if _, exists := smdb.FilterResults[startingPath]; exists {
		newFilterResults[startingPath] = smdb.FilterResults[startingPath]
	} else {
		// If starting path doesn't exist, clear filter results
		smdb.FilterResults = make(map[string]*TreeNode)
		return newFilterResults, nil
	}
	
	// Get and add descendants
	descendants, err := smdb.QueryDescendants(startingPath)
	if err != nil {
		return nil, fmt.Errorf("error querying descendants: %w", err)
	}
	
	for _, item := range descendants {
		if _, exists := smdb.FilterResults[item.Path]; exists {
			newFilterResults[item.Path] = smdb.FilterResults[item.Path]
		}
	}
	
	smdb.FilterResults = newFilterResults
	return newFilterResults, nil
}

// SearchPath searches for rows matching the specified LTREE path expression using operators
func (smdb *SearchMemDB) SearchPath(operator, startingPath string) map[string]*TreeNode {
	// Use the parent class query method
	searchResults := smdb.QueryByOperator(operator, startingPath, "")
	
	newFilterResults := make(map[string]*TreeNode)
	for _, item := range searchResults {
		if _, exists := smdb.FilterResults[item.Path]; exists {
			newFilterResults[item.Path] = smdb.FilterResults[item.Path]
		}
	}
	
	smdb.FilterResults = newFilterResults
	return smdb.FilterResults
}

// FindDescriptions extracts descriptions from all data entries or a specific key
func (smdb *SearchMemDB) FindDescriptions(key interface{}) map[string]string {
	returnValues := make(map[string]string)
	
	// Process all data entries
	for rowKey, rowData := range smdb.data {
		if dataMap, ok := rowData.Data.(map[string]interface{}); ok {
			if description, exists := dataMap["description"]; exists {
				if descStr, ok := description.(string); ok {
					returnValues[rowKey] = descStr
				} else {
					returnValues[rowKey] = ""
				}
			} else {
				returnValues[rowKey] = ""
			}
		} else {
			returnValues[rowKey] = ""
		}
	}
	
	return returnValues
}

// GetFilterResults returns the current filter results
func (smdb *SearchMemDB) GetFilterResults() map[string]*TreeNode {
	// Return a copy to prevent external modification
	results := make(map[string]*TreeNode)
	for key, value := range smdb.FilterResults {
		results[key] = value
	}
	return results
}

// GetFilterResultKeys returns just the keys of current filter results
func (smdb *SearchMemDB) GetFilterResultKeys() []string {
	keys := make([]string, 0, len(smdb.FilterResults))
	for key := range smdb.FilterResults {
		keys = append(keys, key)
	}
	return keys
}

// GetKBs returns all knowledge bases
func (smdb *SearchMemDB) GetKBs() map[string][]string {
	return smdb.kbs
}

// GetLabels returns all labels
func (smdb *SearchMemDB) GetLabels() map[string][]string {
	return smdb.labels
}

// GetNames returns all names
func (smdb *SearchMemDB) GetNames() map[string][]string {
	return smdb.names
}

// GetDecodedKeys returns all decoded keys
func (smdb *SearchMemDB) GetDecodedKeys() map[string][]string {
	return smdb.DecodedKeys
}

